File: src/nacos_mcp_router_typescript/src/nacos_mcp_server_config.ts
import { logger } from './logger';

export interface InputProperty {
  type: string;
  description: string;
}

export class InputPropertyImpl implements InputProperty {
  type: string;
  description: string;

  constructor(type: string, description: string) {
    this.type = type;
    this.description = description;
  }

  static fromDict(data: Record<string, any> | null): InputProperty {
    if (!data || Object.keys(data).length === 0) {
      return new InputPropertyImpl('', '');
    }
    return new InputPropertyImpl(data.type, data.description);
  }
}

export interface InputSchema {
  type: string;
  properties: Record<string, InputProperty>;
}

export class InputSchemaImpl implements InputSchema {
  type: string;
  properties: Record<string, InputProperty>;

  constructor(type: string, properties: Record<string, InputProperty>) {
    this.type = type;
    this.properties = properties;
  }

  static fromDict(data: Record<string, any> | null): InputSchema {
    if (!data || Object.keys(data).length === 0) {
      return new InputSchemaImpl('', {});
    }
    const properties: Record<string, InputProperty> = {};
    for (const [key, value] of Object.entries(data.properties)) {
      properties[key] = InputPropertyImpl.fromDict(value as Record<string, any>);
    }
    return new InputSchemaImpl(data.type, properties);
  }
}

export interface Tool {
  name: string;
  description: string;
  inputSchema: InputSchema;
}

export class ToolImpl implements Tool {
  name: string;
  description: string;
  inputSchema: InputSchema;

  constructor(name: string, description: string, inputSchema: InputSchema) {
    this.name = name;
    this.description = description;
    this.inputSchema = inputSchema;
  }

  static fromDict(data: Record<string, any>): Tool {
    return new ToolImpl(
      data.name,
      data.description,
      InputSchemaImpl.fromDict(data.inputSchema)
    );
  }
}

export interface ToolMeta {
  invokeContext: Record<string, any>;
  enabled: boolean;
  templates: Record<string, string>;
}

export class ToolMetaImpl implements ToolMeta {
  invokeContext: Record<string, any>;
  enabled: boolean;
  templates: Record<string, string>;

  constructor(invokeContext: Record<string, any>, enabled: boolean, templates: Record<string, string>) {
    this.invokeContext = invokeContext;
    this.enabled = enabled;
    this.templates = templates;
  }

  static fromDict(data: Record<string, any>): ToolMeta {
    return new ToolMetaImpl(
      data.invokeContext || {},
      data.enabled ?? true,
      data.templates || {}
    );
  }
}

export interface ToolSpec {
  tools: Tool[];
  toolsMeta: Record<string, ToolMeta>;
}

export class ToolSpecImpl implements ToolSpec {
  tools: Tool[];
  toolsMeta: Record<string, ToolMeta>;

  constructor(tools: Tool[], toolsMeta: Record<string, ToolMeta>) {
    this.tools = tools;
    this.toolsMeta = toolsMeta;
  }

  static fromDict(data: Record<string, any>): ToolSpec {
    return new ToolSpecImpl(
      (data.tools || []).map((t: any) => ToolImpl.fromDict(t)),
      Object.fromEntries(
        Object.entries(data.toolsMeta || {}).map(([k, v]) => [k, ToolMetaImpl.fromDict(v as Record<string, any>)])
      )
    );
  }
}

export interface ServiceRef {
  namespaceId: string;
  groupName: string;
  serviceName: string;
}

export class ServiceRefImpl implements ServiceRef {
  namespaceId: string;
  groupName: string;
  serviceName: string;

  constructor(namespaceId: string, groupName: string, serviceName: string) {
    this.namespaceId = namespaceId;
    this.groupName = groupName;
    this.serviceName = serviceName;
  }

  static fromDict(data: Record<string, any> | null): ServiceRef {
    if (!data || Object.keys(data).length === 0) {
      return new ServiceRefImpl('', '', '');
    }
    return new ServiceRefImpl(
      data.namespaceId,
      data.groupName,
      data.serviceName
    );
  }
}

export interface RemoteServerConfig {
  serviceRef: ServiceRef;
  exportPath: string;
  credentials: Record<string, any>;
}

export class RemoteServerConfigImpl implements RemoteServerConfig {
  serviceRef: ServiceRef;
  exportPath: string;
  credentials: Record<string, any>;

  constructor(serviceRef: ServiceRef, exportPath: string, credentials: Record<string, any>) {
    this.serviceRef = serviceRef;
    this.exportPath = exportPath;
    this.credentials = credentials;
  }

  static fromDict(data: Record<string, any> | null): RemoteServerConfig {
    if (!data || Object.keys(data).length === 0) {
      return new RemoteServerConfigImpl(ServiceRefImpl.fromDict({}), '', {});
    }
    return new RemoteServerConfigImpl(
      ServiceRefImpl.fromDict(data.serviceRef),
      data.exportPath,
      data.credentials || {}
    );
  }
}

export interface BackendEndpoint {
  address: string;
  port: number;
}

export class BackendEndpointImpl implements BackendEndpoint {
  address: string;
  port: number;

  constructor(address: string, port: number) {
    this.address = address;
    this.port = port;
  }

  static fromDict(data: Record<string, any> | null): BackendEndpoint {
    if (!data || Object.keys(data).length === 0) {
      return new BackendEndpointImpl('', -1);
    }
    return new BackendEndpointImpl(data.address, data.port);
  }
}

export interface NacosMcpServerConfig {
  name: string;
  protocol: string;
  description: string | null;
  version: string;
  remoteServerConfig: RemoteServerConfig;
  localServerConfig: Record<string, any>;
  enabled: boolean;
  capabilities: string[];
  backendEndpoints: BackendEndpoint[];
  toolSpec: ToolSpec;
  getToolDescription(): string;
}

export class NacosMcpServerConfigImpl implements NacosMcpServerConfig {
  name: string;
  protocol: string;
  description: string | null;
  version: string;
  remoteServerConfig: RemoteServerConfig;
  localServerConfig: Record<string, any>;
  enabled: boolean;
  capabilities: string[];
  backendEndpoints: BackendEndpoint[];
  toolSpec: ToolSpec;

  constructor(
    name: string,
    protocol: string,
    description: string | null,
    version: string,
    remoteServerConfig: RemoteServerConfig,
    localServerConfig: Record<string, any>,
    enabled: boolean,
    capabilities: string[],
    backendEndpoints: BackendEndpoint[],
    toolSpec: ToolSpec
  ) {
    this.name = name;
    this.protocol = protocol;
    this.description = description;
    this.version = version;
    this.remoteServerConfig = remoteServerConfig;
    this.localServerConfig = localServerConfig;
    this.enabled = enabled;
    this.capabilities = capabilities;
    this.backendEndpoints = backendEndpoints;
    this.toolSpec = toolSpec;
  }

  static fromDict(data: Record<string, any>): NacosMcpServerConfig {
    const toolSpecData = data.toolSpec;
    const backendEndpointsData = data.backendEndpoints;

    try {
      return new NacosMcpServerConfigImpl(
        data.name,
        data.protocol,
        data.description,
        data.version,
        RemoteServerConfigImpl.fromDict(data.remoteServerConfig),
        data.localServerConfig || {},
        data.enabled ?? true,
        data.capabilities || [],
        backendEndpointsData ? backendEndpointsData.map((e: any) => BackendEndpointImpl.fromDict(e)) : [],
        toolSpecData ? ToolSpecImpl.fromDict(toolSpecData) : new ToolSpecImpl([], {})
      );
    } catch (error) {
      logger.warn(`failed to parse NacosMcpServerConfig from data: ${JSON.stringify(data)}`, error);
      throw new Error('failed to parse NacosMcpServerConfig from data');
    }
  }

  static fromString(string: string): NacosMcpServerConfig {
    return NacosMcpServerConfigImpl.fromDict(JSON.parse(string));
  }

  getToolDescription(): string {
    let des = this.description || '';
    for (const tool of this.toolSpec.tools) {
      if (tool.description) {
        des += '\n' + tool.description;
      }
    }
    return des;
  }
}

File: src/python/src/nacos_mcp_router/nacos_mcp_server_config.py
#-*- coding: utf-8 -*-
import json
from dataclasses import dataclass, field
from typing import List, Dict, Any
from .logger import NacosMcpRouteLogger


@dataclass
class InputProperty:
    type: str
    description: str

    @classmethod
    def from_dict(cls, data: dict) -> "InputProperty":
        if data is None or len(data) == 0:
            return InputProperty(type="", description="")
        return cls(
            type=data["type"],
            description=data["description"]
        )

@dataclass
class InputSchema:
    type: str
    properties: Dict[str, InputProperty]

    @classmethod
    def from_dict(cls, data: dict) -> "InputSchema":
        if data is None or len(data) == 0:
            return InputSchema(type="", properties={})
        return cls(
            type=data["type"],
            properties={k: InputProperty.from_dict(v) for k, v in data["properties"].items()}
        )

@dataclass
class Tool:
    name: str
    description: str
    input_schema: dict

    @classmethod
    def from_dict(cls, data: dict) -> "Tool":
        return cls(
            name=data["name"],
            description=data["description"],
            input_schema=data["inputSchema"]
        )

@dataclass
class ToolMeta:
    invoke_context: Dict[str, Any]
    enabled: bool
    templates: Dict[str, str]

    @classmethod
    def from_dict(cls, data: dict) -> "ToolMeta":
        return cls(
            invoke_context=data.get("invokeContext", {}),
            enabled=data.get("enabled", True),
            templates=data.get("templates", {})
        )

@dataclass
class ToolSpec:
    tools: List[Tool]
    tools_meta: Dict[str, ToolMeta]

    @classmethod
    def from_dict(cls, data: dict) -> "ToolSpec":
        return cls(
            tools=[Tool.from_dict(t) for t in data.get("tools", [])],
            tools_meta={k: ToolMeta.from_dict(v) for k, v in data.get("toolsMeta", {}).items()}
        )

# ------------------ 主结构 ------------------
@dataclass
class ServiceRef:
    namespace_id: str
    group_name: str
    service_name: str

    @classmethod
    def from_dict(cls, data: dict) -> "ServiceRef":
        if data is None or len(data) == 0:
            return ServiceRef(namespace_id="", group_name="", service_name="")
        return cls(
            namespace_id=data["namespaceId"],
            group_name=data["groupName"],
            service_name=data["serviceName"]
        )

@dataclass
class RemoteServerConfig:
    service_ref: ServiceRef
    export_path: str
    credentials: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: dict) -> "RemoteServerConfig":
        if data is None or len(data) == 0:
            return RemoteServerConfig(service_ref=ServiceRef.from_dict({}), export_path="", credentials={})
        return cls(
            service_ref=ServiceRef.from_dict(data["serviceRef"]),
            export_path=data["exportPath"],
            credentials=data.get("credentials", {})
        )

@dataclass
class BackendEndpoint:
    address: str
    port: int

    @classmethod
    def from_dict(cls, data: dict) -> "BackendEndpoint":
        if data is None or len(data) == 0:
            return BackendEndpoint(address="", port=-1)
        return cls(
            address=data["address"],
            port=data["port"]
        )

@dataclass
class NacosMcpServerConfig:
    name: str
    protocol: str
    description: str | None
    version: str
    id: str | None
    remote_server_config: RemoteServerConfig
    local_server_config: Dict[str, Any] = field(default_factory=dict)
    enabled: bool = True
    capabilities: List[str] = field(default_factory=list)
    backend_endpoints: List[BackendEndpoint] = field(default_factory=list)
    tool_spec: ToolSpec = field(default_factory=lambda: ToolSpec(tools=[], tools_meta={}))

    @classmethod
    def from_dict(cls, data: dict) -> "NacosMcpServerConfig":
        tool_spec_data = data.get("toolSpec")
        backend_endpoints_data = data.get("backendEndpoints")
        try:
            return cls(
                name=data["name"],
                protocol=data["protocol"],
                description=data["description"],
                version=data["version"],
                remote_server_config=RemoteServerConfig.from_dict(data["remoteServerConfig"]),
                local_server_config=data.get("localServerConfig", {}) if data.get("localServerConfig") else {},
                enabled=data.get("enabled", True),
                capabilities=data.get("capabilities", []),
                backend_endpoints=[BackendEndpoint.from_dict(e) for e in data.get("backendEndpoints", [])] if backend_endpoints_data else [],
                tool_spec=ToolSpec.from_dict(tool_spec_data) if tool_spec_data else ToolSpec(tools=[], tools_meta={}),
                id=data["id"] if data.get("id") else None
            )
        except Exception as e:
            NacosMcpRouteLogger.get_logger().warning("failed to parse NacosMcpServerConfig from data: %s", data,  exc_info=e)
            raise Exception("failed to parse NacosMcpServerConfig from data")

    @classmethod
    def from_string(cls, string: str) -> "NacosMcpServerConfig":
        return cls.from_dict(json.loads(string))

    def get_tool_description(self) -> str:
        des = "" if self.description is None else self.description
        for tool in self.tool_spec.tools:
            if tool.description is not None:
                des += "\n" + tool.description

        return des
